### ISSUE TITLE
File::find() fails

### ISSUE THREAD

**Issue by phptek** [silverstripe/silverstripe-framework#4076]
The `File::find()` method, fails to yield a file I have uploaded. Looking at the DB it has a ParentID of 30, however the `find()` method assumes the ParentID is always "0" for some reason.

I have grepped the filesystem for uses of this method in both `framework` and `cms`; It's used once both: 

In `cms` it's ultimately called by an unused method `SiteTreeLinkTracking::augmentSyncLinkTracking()`. In `framework` it's used in `HtmlEditorField::saveInto()` which I imagine _is_ used so am confused why this hasn't been spotted before.

Here's the relevant logic:

    		$parentID = 0;
		$item = null;
		foreach($parts as $part) {
			if($part == ASSETS_DIR && !$parentID) continue;
			$SQL_part = Convert::raw2sql($part);
			$item = DataObject::get_one("File", "\"Name\" = '$SQL_part' AND \"ParentID\" = $parentID");
			if(!$item) break;
    $parentID = $item->ID;
    ...

If the uploaded file has a `$ParentID` of anything other than 0, then I don't see how the call to `get_one()` could ever work, as $ParentID is zero by default and if the query fails (which it will in my case) then that variable can never be reset because the loop is broken out of immediately after if the query fails by returning null.

Tell me I'm missing something :-)

**Comment by phptek** [comment]
Modifying that method slightly makes it work, but requires x2 ORM queries to fetch a single file:

                if($parentID >0) {
                $item = DataObject::get_one("File", "\"Name\" = '$SQL_part' AND \"ParentID\" = $parentID");
            } else {
                $item = DataObject::get_one("File", "\"Name\" = '$SQL_part'");
            }

..however - stupid question perhaps: Why does the query require a ParentID anyway? Filenames are unique in the CMS, so can we just lose that SQL clause entirely?

**Comment by chillu** [comment]
Hey Russ! `File::find()` accepts a file *path* rather than just a file *name*, hence needs to check the filesystem hierarchy (as mirrored by `ParentID` in the `File` table). It does this by using `$parts = explode("/", $filename);`, a crucial bit which you've left out of the logic you've posted. For completeness, here's `File::find()` as implemented in 3.1:

```php
	/**
	 * Find a File object by the given filename.
	 * 
	 * @param String $filename Matched against the "Name" property.
	 * @return mixed null if not found, File object of found file
	 */
	public static function find($filename) {
		// Get the base file if $filename points to a resampled file
		$filename = preg_replace('/_resampled\/[^-]+-/', '', $filename);

		// Split to folders and the actual filename, and traverse the structure.
		$parts = explode("/", $filename);
		$parentID = 0;
		$item = null;
		foreach($parts as $part) {
			if($part == ASSETS_DIR && !$parentID) continue;
			$SQL_part = Convert::raw2sql($part);
			$item = DataObject::get_one("File", "\"Name\" = '$SQL_part' AND \"ParentID\" = $parentID");
			if(!$item) break;
			$parentID = $item->ID;
		}
		
		return $item;
	}
```

So its behaving as expected. If you want to get a file by name regardless of its folder (and accept the danger of getting a different file if a duplicate exists with the same name in a different folder), simply use `File::get()`.
